java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileWriter.java

```diff
@@ -155,31 +143,37 @@ private void writeBatch(VectorSchemaRoot recordBatch) throws IOException {
       generator.writeArrayFieldStart("columns");
       for (Field field : recordBatch.getSchema().getFields()) {
         FieldVector vector = recordBatch.getVector(field.getName());
-        writeVector(field, vector);
+        writeFromVectorIntoJson(field, vector);
       }
       generator.writeEndArray();
     }
     generator.writeEndObject();
   }
 
-  private void writeVector(Field field, FieldVector vector) throws IOException {
+  private void writeFromVectorIntoJson(Field field, FieldVector vector) throws IOException {
     List<ArrowVectorType> vectorTypes = field.getTypeLayout().getVectorTypes();
-    List<BufferBacked> fieldInnerVectors = vector.getFieldInnerVectors();
-    if (vectorTypes.size() != fieldInnerVectors.size()) {
-      throw new IllegalArgumentException("vector types and inner vectors are not the same size: " + vectorTypes.size() + " != " + fieldInnerVectors.size());
+    List<ArrowBuf> vectorBuffers = vector.getFieldBuffers();
+    if (vectorTypes.size() != vectorBuffers.size()) {
+      throw new IllegalArgumentException("vector types and inner vector buffers are not the same size: " + vectorTypes.size() + " != " + vectorBuffers.size());
     }
     generator.writeStartObject();
     {
       generator.writeObjectField("name", field.getName());
-      int valueCount = vector.getAccessor().getValueCount();
+      int valueCount = vector.getValueCount();
       generator.writeObjectField("count", valueCount);
       for (int v = 0; v < vectorTypes.size(); v++) {
         ArrowVectorType vectorType = vectorTypes.get(v);
-        BufferBacked innerVector = fieldInnerVectors.get(v);
+        ArrowBuf vectorBuffer = vectorBuffers.get(v);
         generator.writeArrayFieldStart(vectorType.getName());
-        ValueVector valueVector = (ValueVector) innerVector;
-        for (int i = 0; i < valueVector.getAccessor().getValueCount(); i++) {
-          writeValueToGenerator(valueVector, i);
+        final int bufferValueCount = (vectorType.equals(OFFSET)) ? valueCount + 1 : valueCount;
+        for (int i = 0; i < bufferValueCount; i++) {
+          if (vectorType.equals(DATA) &&
+                  (vector.getMinorType() == Types.MinorType.VARCHAR ||
+                          vector.getMinorType() == Types.MinorType.VARBINARY)) {
```

Comment:
style: indentation seems off in this if statement

Reply:
Done.